contracts/libs/data-structures/AvlTree.sol

```diff
@@ -0,0 +1,823 @@
+     */
+    function getMin(AddressAVL storage tree) internal view returns (address) {
+        return _asAddress(_getMin(tree._tree.tree, tree._tree.root));
+    }
+
+    /**
+     * @notice The function to retrieve the maximum key in the address tree.
+     * Complexity is O(log(n)), where n is the number of elements in the tree.
+     *
+     * @param tree self.
+     * @return The maximum key in the tree.
+     */
+    function getMax(AddressAVL storage tree) internal view returns (address) {
+        return _asAddress(_getMax(tree._tree.tree, tree._tree.root));
+    }
+
+    /**
+     * @notice The function to return the key of the root element of the address tree.
+     * Complexity is O(1).
+     *
+     * @param tree self.
+     * @return The key of the root element of the uint256 tree.
+     */
+    function root(AddressAVL storage tree) internal view returns (address) {
+        return _asAddress(tree._tree.root);
+    }
+
+    /**
+     * @notice The function to retrieve the size of the address tree.
+     * @param tree self.
+     * @return The size of the tree.
+     */
+    function treeSize(AddressAVL storage tree) internal view returns (uint256) {
+        return tree._tree.treeSize;
+    }
+
+    /**
+     * @notice The function to perform an in-order traversal of the address tree.
+     * @param tree self.
+     * @return An array of keys in in-order traversal.
+     */
+    function inOrderTraversal(AddressAVL storage tree) internal view returns (address[] memory) {
+        return _getTraversal(tree._tree, _inOrderTraversal).asAddressArray();
+    }
+
+    /**
+     * @notice The function to perform an pre-order traversal of the address tree.
+     * @param tree self.
+     * @return An array of keys in pre-order traversal.
+     */
+    function preOrderTraversal(AddressAVL storage tree) internal view returns (address[] memory) {
+        return _getTraversal(tree._tree, _preOrderTraversal).asAddressArray();
+    }
+
+    /**
+     * @notice The function to perform an post-order traversal of the address tree.
+     * @param tree self.
+     * @return An array of keys in post-order traversal.
+     */
+    function postOrderTraversal(AddressAVL storage tree) internal view returns (address[] memory) {
+        return _getTraversal(tree._tree, _postOrderTraversal).asAddressArray();
+    }
+
+    /**
+     * @notice The function to check whether the custom comparator function is set for the address tree.
+     * @param tree self.
+     * @return True if the custom comparator function is set, false otherwise.
+     */
+    function isCustomComparatorSet(AddressAVL storage tree) internal view returns (bool) {
+        return tree._tree.isCustomComparatorSet;
+    }
+
+    /**
+     *************************
+     *     Internal Tree     *
+     *************************
+     */
+
+    struct Node {
+        bytes32 key;
+        bytes value;
+        uint256 height;
+        bytes32 left;
+        bytes32 right;
```

Comment:
```suggestion
        bytes32 key;
        bytes32 value;
        uint64 height;
        uint64 parent;
        uint64 left;
        uint64 right;
```

And store the nodes with incremental indexes. Yep, the `search` function will be logarithmic but we 

1. Save storage and gas
2. Able to implement `iterator` design pattern to traverse the tree